File: clm-js/src/document.ts
/**
 * Coarse-grained parser for the Claude Memory Format.
 *
 * Supports both CLM/1.0 (Unicode brackets `⟦NAME⟧`) and CLM/2.x+/3.0
 * (ASCII brackets `[NAME]`). The contract is round-trip:
 * `Document.parse(text).toString() === text` byte-for-byte for any document
 * accepted by the grammar.
 *
 * v3.0 trim-aware validation lives in `./validate.ts`.
 */

const SEC_OPEN = "\u27E6"; // ⟦
const SEC_CLOSE = "\u27E7"; // ⟧

export class ParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ParseError";
  }
}

export class MutationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MutationError";
  }
}

export interface Section {
  name: string;
  openLine: string;
  body: string[];
  closeLine: string;
}

export class Document {
  header: string[] = [];
  triviaAfterHeader: string[] = [];
  /** Each section paired with the trivia (typically blank lines) that follows it. */
  sections: Array<{ section: Section; trivia: string[] }> = [];
  closer: string[] = [];
  trailingNewline = true;

  // ---- parsing ----

  static parse(text: string): Document {
    const doc = new Document();
    doc.trailingNewline = text.endsWith("\n");
    const body = doc.trailingNewline ? text.slice(0, -1) : text;
    const lines = body.length === 0 ? [] : body.split("\n");

    let idx = 0;

    // Header
    if (idx >= lines.length || !isTripleSemi(lines[idx]!)) {
      throw new ParseError("missing file header (expected lines starting with ';;;')");
    }
    let foundTerminator = false;
    while (idx < lines.length && isTripleSemi(lines[idx]!)) {
      doc.header.push(lines[idx]!);
      if (isHeaderTerminator(lines[idx]!)) {
        idx++;
        foundTerminator = true;
        break;
      }
      idx++;
    }
    if (!foundTerminator) {
      throw new ParseError("missing header terminator (expected ';;; ---')");
    }

    // Trivia after header (blank lines)
    while (idx < lines.length && isBlank(lines[idx]!)) {
      doc.triviaAfterHeader.push(lines[idx]!);
      idx++;
    }

    // Sections
    while (true) {
      if (idx >= lines.length) {
        throw new ParseError("missing file closer (expected ';;; EOF ...')");
      }
      if (isCloserStart(lines[idx]!)) {
        break;
      }
      const openLine = lines[idx]!;
      const name = parseSectionOpen(openLine);
      if (name === null) {
        throw new ParseError(
          `unexpected line at top level (line ${idx + 1}): ${JSON.stringify(openLine)}`
        );
      }
      const openLineIdx = idx;
      idx++;

      const sectionBody: string[] = [];
      let closeLine: string | null = null;
      while (idx < lines.length) {
        if (isSectionClose(lines[idx]!)) {
          closeLine = lines[idx]!;
          idx++;
          break;
        }
        if (parseSectionOpen(lines[idx]!) !== null) {
          throw new ParseError(`nested section opens are not allowed (line ${idx + 1})`);
        }
        sectionBody.push(lines[idx]!);
        idx++;
      }
      if (closeLine === null) {
        throw new ParseError(
          `section [${name}] opened at line ${openLineIdx + 1} was never closed (';;')`
        );
      }

      const trivia: string[] = [];
      while (idx < lines.length && isBlank(lines[idx]!)) {
        trivia.push(lines[idx]!);
        idx++;
      }

      doc.sections.push({
        section: { name, openLine, body: sectionBody, closeLine },
        trivia,
      });
    }

    // Closer
    while (idx < lines.length) {
      if (!isTripleSemi(lines[idx]!)) {
        throw new ParseError(`unexpected content after file closer (line ${idx + 1})`);
      }
      doc.closer.push(lines[idx]!);
      idx++;
    }
    if (doc.closer.length === 0 || !isCloserStart(doc.closer[0]!)) {
      throw new ParseError("missing file closer (expected ';;; EOF ...')");
    }

    return doc;
  }

  // ---- accessors ----

  section(name: string): Section | undefined {
    const found = this.sections.find((entry) => entry.section.name === name);
    return found?.section;
  }

  // ---- mutation ----

  appendToSection(name: string, text: string): void {
    const entry = this.sections.find((e) => e.section.name === name);
    if (!entry) throw new MutationError(`no such section: [${name}]`);
    for (const line of text.split("\n")) {
      entry.section.body.push(line);
    }
  }

  appendSignature(line: string): void {
    this.closer.push(line);
  }

  // ---- serialization ----

  toString(): string {
    const out: string[] = [];
    for (const line of this.header) {
      out.push(line, "\n");
    }
    for (const line of this.triviaAfterHeader) {
      out.push(line, "\n");
    }
    for (const { section, trivia } of this.sections) {
      out.push(section.openLine, "\n");
      for (const bodyLine of section.body) {
        out.push(bodyLine, "\n");
      }
      out.push(section.closeLine, "\n");
      for (const line of trivia) {
        out.push(line, "\n");
      }
    }
    const last = this.closer.length - 1;
    for (let i = 0; i < this.closer.length; i++) {
      out.push(this.closer[i]!);
      if (i < last || this.trailingNewline) {
        out.push("\n");
      }
    }
    return out.join("");
  }
}

// ---- line classifiers ----

function isTripleSemi(line: string): boolean {
  return line.startsWith(";;;");
}

function isHeaderTerminator(line: string): boolean {
  if (!line.startsWith(";;;")) return false;
  return line.slice(3).trim() === "---";
}

function isCloserStart(line: string): boolean {
  if (!line.startsWith(";;;")) return false;
  // Unicode-equivalent whitespace strip — Rust uses .trim_start() and Python .lstrip(),
  // both of which handle full Unicode whitespace. Limiting to [\t ] would reject docs
  // with NBSP (U+00A0) or other Unicode whitespace that the other implementations accept.
  const after = line.slice(3).replace(/^\s+/, "");
  if (!after.startsWith("EOF")) return false;
  const rest = after.slice(3);
  if (rest.length === 0) return true;
  const c = rest.charCodeAt(0);
  // Not ASCII alphanumeric.
  return !((c >= 0x30 && c <= 0x39) || (c >= 0x41 && c <= 0x5a) || (c >= 0x61 && c <= 0x7a));
}

function isBlank(line: string): boolean {
  if (line.length === 0) return true;
  return /^\s*$/.test(line);
}

function isSectionClose(line: string): boolean {
  return line.replace(/\s+$/, "") === ";;";
}

function parseSectionOpen(line: string): string | null {
  const trimmed = line.replace(/\s+$/, "");
  // CLM/1.0: Unicode brackets
  if (trimmed.startsWith(SEC_OPEN) && trimmed.endsWith(SEC_CLOSE)) {
    return validateSectionName(trimmed.slice(SEC_OPEN.length, -SEC_CLOSE.length));
  }
  // CLM/2.x+/3.0: ASCII brackets
  if (trimmed.startsWith("[") && trimmed.endsWith("]")) {
    return validateSectionName(trimmed.slice(1, -1));
  }
  return null;
}

/**
 * Permissive parser-level grammar; strict semantics enforced in `./validate.ts`.
 *
 * Plain section names: `[A-Z][A-Z0-9.]*` (uppercase only).
 * `DELTA.<suffix>`: parser accepts any identifier-shaped suffix; validator enforces
 * the spec session-id grammar so malformed names surface as InvalidDeltaSessionId.
 */
function validateSectionName(inner: string): string | null {
  if (inner.length === 0) return null;
  // Permissive parse for DELTA.<suffix>.
  if (inner.startsWith("DELTA.")) {
    const suffix = inner.slice("DELTA.".length);
    if (suffix.length === 0) return null;
    if (!/^[A-Za-z0-9._-]+$/.test(suffix)) return null;
    return inner;
  }
  // Plain section name: strict.
  if (!/^[A-Z][A-Z0-9.]*$/.test(inner)) return null;
  return inner;
}

File: clm-py/src/clm/document.py
"""Coarse-grained parser for the Claude Memory Format.

Supports both CLM/1.0 (Unicode brackets `⟦NAME⟧`) and CLM/2.x+/3.0
(ASCII brackets `[NAME]`). The contract is round-trip:
`str(Document.parse(text)) == text` byte-for-byte for any document
accepted by the grammar.

v3.0 trim-aware validation lives in :mod:`clm.validate`.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable, Optional


SEC_OPEN = "\u27E6"   # ⟦
SEC_CLOSE = "\u27E7"  # ⟧


class ParseError(Exception):
    """A document failed to parse against the CLM grammar."""


class MutationError(Exception):
    """A document mutation (e.g. append_to_section) targeted a missing section."""


@dataclass
class Section:
    """A `[NAME] ... ;;` section in a CLM document."""

    name: str
    open_line: str
    body: list[str] = field(default_factory=list)
    close_line: str = ";;"


@dataclass
class Document:
    """A parsed CLM document. Round-trips byte-identically via ``str(doc)``."""

    header: list[str] = field(default_factory=list)
    trivia_after_header: list[str] = field(default_factory=list)
    # Each section is paired with the trivia (typically blank lines) that follows it.
    sections: list[tuple[Section, list[str]]] = field(default_factory=list)
    closer: list[str] = field(default_factory=list)
    trailing_newline: bool = True

    # ---- parsing ----

    @classmethod
    def parse(cls, text: str) -> "Document":
        trailing_newline = text.endswith("\n")
        body = text[:-1] if trailing_newline else text
        lines = body.split("\n") if body else []

        idx = 0

        # Header
        if idx >= len(lines) or not _is_triple_semi(lines[idx]):
            raise ParseError("missing file header (expected lines starting with ';;;')")
        header: list[str] = []
        found_terminator = False
        while idx < len(lines) and _is_triple_semi(lines[idx]):
            header.append(lines[idx])
            if _is_header_terminator(lines[idx]):
                idx += 1
                found_terminator = True
                break
            idx += 1
        if not found_terminator:
            raise ParseError("missing header terminator (expected ';;; ---')")

        # Trivia after header (blank lines)
        trivia_after_header: list[str] = []
        while idx < len(lines) and _is_blank(lines[idx]):
            trivia_after_header.append(lines[idx])
            idx += 1

        # Sections
        sections: list[tuple[Section, list[str]]] = []
        while True:
            if idx >= len(lines):
                raise ParseError("missing file closer (expected ';;; EOF ...')")
            if _is_closer_start(lines[idx]):
                break
            open_line = lines[idx]
            name = _parse_section_open(open_line)
            if name is None:
                raise ParseError(f"unexpected line at top level (line {idx + 1}): {open_line!r}")
            open_line_idx = idx
            idx += 1

            section_body: list[str] = []
            close_line: Optional[str] = None
            while idx < len(lines):
                if _is_section_close(lines[idx]):
                    close_line = lines[idx]
                    idx += 1
                    break
                if _parse_section_open(lines[idx]) is not None:
                    raise ParseError(
                        f"nested section opens are not allowed (line {idx + 1})"
                    )
                section_body.append(lines[idx])
                idx += 1
            if close_line is None:
                raise ParseError(
                    f"section [{name}] opened at line {open_line_idx + 1} was never closed (';;')"
                )

            trivia: list[str] = []
            while idx < len(lines) and _is_blank(lines[idx]):
                trivia.append(lines[idx])
                idx += 1

            sections.append(
                (
                    Section(
                        name=name,
                        open_line=open_line,
                        body=section_body,
                        close_line=close_line,
                    ),
                    trivia,
                )
            )

        # Closer
        closer: list[str] = []
        while idx < len(lines):
            if not _is_triple_semi(lines[idx]):
                raise ParseError(f"unexpected content after file closer (line {idx + 1})")
            closer.append(lines[idx])
            idx += 1
        if not closer or not _is_closer_start(closer[0]):
            raise ParseError("missing file closer (expected ';;; EOF ...')")

        return cls(
            header=header,
            trivia_after_header=trivia_after_header,
            sections=sections,
            closer=closer,
            trailing_newline=trailing_newline,
        )

    # ---- accessors ----

    def section(self, name: str) -> Optional[Section]:
        for sec, _ in self.sections:
            if sec.name == name:
                return sec
        return None

    # ---- mutation ----

    def append_to_section(self, name: str, text: str) -> None:
        for sec, _ in self.sections:
            if sec.name == name:
                sec.body.extend(text.split("\n"))
                return
        raise MutationError(f"no such section: [{name}]")

    def append_signature(self, line: str) -> None:
        self.closer.append(line)

    # ---- serialization ----

    def __str__(self) -> str:
        out: list[str] = []
        for line in self.header:
            out.append(line)
            out.append("\n")
        for line in self.trivia_after_header:
            out.append(line)
            out.append("\n")
        for sec, trivia in self.sections:
            out.append(sec.open_line)
            out.append("\n")
            for body_line in sec.body:
                out.append(body_line)
                out.append("\n")
            out.append(sec.close_line)
            out.append("\n")
            for line in trivia:
                out.append(line)
                out.append("\n")
        last = len(self.closer) - 1
        for i, line in enumerate(self.closer):
            out.append(line)
            if i < last or self.trailing_newline:
                out.append("\n")
        return "".join(out)


# ---- line classifiers ----

def _is_triple_semi(line: str) -> bool:
    return line.startswith(";;;")


def _is_header_terminator(line: str) -> bool:
    if not line.startswith(";;;"):
        return False
    return line[3:].strip() == "---"


def _is_closer_start(line: str) -> bool:
    if not line.startswith(";;;"):
        return False
    after = line[3:].lstrip()
    if not after.startswith("EOF"):
        return False
    rest = after[3:]
    if not rest:
        return True
    return not (rest[0].isascii() and rest[0].isalnum())


def _is_blank(line: str) -> bool:
    return all(c.isspace() for c in line) if line else True


def _is_section_close(line: str) -> bool:
    return line.rstrip() == ";;"


def _parse_section_open(line: str) -> Optional[str]:
    """Recognize both Unicode (CLM/1.0) and ASCII (CLM/2.x+/3.0) section opens."""
    trimmed = line.rstrip()
    # CLM/1.0: Unicode brackets
    if trimmed.startswith(SEC_OPEN) and trimmed.endswith(SEC_CLOSE):
        return _validate_section_name(trimmed[len(SEC_OPEN):-len(SEC_CLOSE)])
    # CLM/2.x+/3.0: ASCII brackets
    if trimmed.startswith("[") and trimmed.endswith("]"):
        return _validate_section_name(trimmed[1:-1])
    return None


def _validate_section_name(inner: str) -> Optional[str]:
    """Permissive parser-level grammar; strict semantics enforced in :mod:`clm.validate`.

    Plain section names: ``[A-Z][A-Z0-9.]*`` (uppercase only).
    DELTA.<suffix>: parser accepts any identifier-shaped suffix; validator enforces
    the spec session-id grammar so malformed names surface as InvalidDeltaSessionId.
    """
    if not inner:
        return None
    # Permissive parse for DELTA.<suffix>.
    if inner.startswith("DELTA."):
        suffix = inner[len("DELTA."):]
        if not suffix:
            return None
        if not all(c.isascii() and (c.isalnum() or c in "._-") for c in suffix):
            return None
        return inner
    # Plain section name: strict.
    first = inner[0]
    if not (first.isascii() and first.isupper()):
        return None
    if not all(c.isascii() and (c.isupper() or c.isdigit() or c == ".") for c in inner[1:]):
        return None
    return inner


__all__ = [
    "Document",
    "Section",
    "ParseError",
    "MutationError",
]
